miniapps/autodiff/par_example.cpp

```diff
@@ -0,0 +1,521 @@
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+//
+//                       MFEM AD Example  - Parallel Version
+//
+// Compile with: make par_example
+//
+// Sample runs:
+//   mpirun -np 2 paradiff -m ../data/beam-quad.mesh -pp 3.8
+//   mpirun -np 2 paradiff -m ../data/beam-tri.mesh  -pp 7.2
+//   mpirun -np 2 paradiff -m ../data/beam-hex.mesh
+//   mpirun -np 2 paradiff -m ../data/beam-tet.mesh
+//   mpirun -np 2 paradiff -m ../data/beam-wedge.mesh
+//
+// Description:  This examples solves a quasi-static nonlinear
+//               p-Laplacian problem with zero Dirichlet boundary
+//               conditions applied on all defined boundaries
+//
+//           The example demonstrates the use of nonlinear operators combined
+//           with automatic differentiation (AD). The definitions of the
+//           integrators are written in the example.hpp.  Selecting integrator=0
+//           will use the manually implemented integrator.  Selecting
+//           integrator=1,2 will utilize one of the AD integrators.
+//
+//        We recommend viewing examples 1 and 19, before playing with this
+//        example.
+
+#include "example.hpp"
+
+using namespace mfem;
+
+///Non-linear solver for the p-Laplacian problem.
+class ParNLSolverPLaplacian
+{
+public:
+   ///Constructor Input: imesh - FE mesh, finite element space,
+   /// power for the p-Laplacian, external load (source, input),
+   /// regularization parameter
+   ParNLSolverPLaplacian(MPI_Comm comm, ParMesh& imesh,
+                         ParFiniteElementSpace& ifespace,
+                         double powerp=2,
+                         Coefficient* load=nullptr,
+                         double regularizationp=1e-7)
+   {
+      lcomm = comm;
+
+      //default parameters for
+      //the Newton solver
+      newton_rtol = 1e-4;
+      newton_atol = 1e-8;
+      newton_iter = 10;
+
+      //linear solver
+      linear_rtol = 1e-7;
+      linear_atol = 1e-15;
+      linear_iter = 500;
+
+      print_level = 0;
+
+      //set the mesh
+      mesh=&imesh;
+
+      //set the fespace
+      fespace=&ifespace;
+
+      //set the parameters
+      plap_epsilon=new ConstantCoefficient(regularizationp);
+      plap_power=new ConstantCoefficient(powerp);
+      if (load==nullptr)
+      {
+         plap_input=new ConstantCoefficient(1.0);
+         input_ownership=true;
+      }
+      else
+      {
+         plap_input=load;
+         input_ownership=false;
+      }
+
+      nf=nullptr;
+      ns=nullptr;
+      gmres=nullptr;
+      prec=nullptr;
+
+      //set the default integrator
+      integ=0; //hand coded
+   }
+
+   ~ParNLSolverPLaplacian()
+   {
+      if (nf!=nullptr) { delete nf;}
```

Comment:
I believe the C++ standard says it's okay to delete a nullptr. Hence you can just say `delete nf;` without the `if`.

Reply:
modified